# Exploring Hacker News Posts

<p>In this project we will work with a data set of submission to the popular technology site, Hacker News.</p>

</p>Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.</p>

<u> NOTE: The data set used in this project can be found in this <strong>link</strong></u>: <a href = "https://www.kaggle.com/hacker-news/hacker-news-posts"> Click Here </a>

We're specifically interested in posts whose titles begin with either Ask HN or Show HN. Users submit Ask HN posts to ask the Hacker News community a specific question. Below are a couple examples:

Ask HN: How to improve my personal website?

Ask HN: Am I the only one outraged by Twitter shutting down share counts?

Ask HN: Aby recent changes to CSS that broke mobile?

Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting. Below are a couple of examples:

Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform'

Show HN: Something pointless I made

Show HN: Shanhu.io, a programming playground powered by e8vm

We'll compare these two types of posts to determine the following:

<ul>Do Ask HN or Show HN receive more comments on average?</ul>
<ul>Do posts created at a certain time receive more comments on average?</ul>

## Introduction:
Let's start by importing the libraries we need and reading the data set into a list of lists.



In [3]:
opened_file = open('hacker_news.csv')
from csv import reader
read_file = reader(opened_file)
hn = list(read_file)


#Printing the first 5  rows, gives us the following: 
print(hn[0])
print(hn[1])
print(hn[2])
print(hn[3])
print(hn[4])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52']
['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30']
['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20']
['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']


<p>Notice that the first list in the inner lists contains the column headers, and the lists after contain the data for one row. In order to analyze our data, we need to first remove the row containing the column headers. Let's remove that first row next.</p>

In [6]:
headers = hn[0]  #Assign headers to first row
hn.remove(hn[0]) #Remove first row from hn
print(headers)   #Display headers
print(hn[0])     #As well as the first 5 rows of hn
print(hn[1])
print(hn[2])
print(hn[3])
print(hn[4])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52']
['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30']
['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20']
['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']
['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']


<p>Now that we've removed the headers from hn, we're ready to filter our data. Since we're only concerned with post titles beginning with Ask HN or Show HN, we'll create new lists of lists containing just the data for those titles.</p>

<p>To find the posts that begin with either Ask HN or Show HN, we'll use the string method startswith. Given a string object, say, string1, we can check if starts with, say, dq by inspecting the output of the object string1.startswith('dq'). If string1 starts with dq, it will return True, otherwise it will return False.</p>

<code>
print('dataquest'.startswith('Data'))

print('dataquest'.startswith('data'))

False

True

</code>
<p>
In the example above, the first print call gives us False because dataquest does not start with Data. The second print call prints True because dataquest does start with data. Capitalization matters.

If we wish to control for case, we can use the lower method which returns a lowercase version of the starting string. Here's an example:</p>

<code>
print('DataQuest'.lower())

dataquest
</code>

Let's use these methods to separate posts beginning with Ask HN and Show HN (and case variations) into two different lists next.

In [8]:
#Creating three empty lists:

ask_posts   = []
show_posts  = []
other_posts = []

#Loop through each row
for row in hn:
    title = row[1] #Get title from the data set (index 1)
    
    
    #If the lowercase version of title starts with ask hn, 
    #append the row to ask_posts.
    #Else if the lowercase version of title starts with 
    #show hn, append the row to show_posts.
    #Otherwise, append to other_posts.
 
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
#Displaying the number of posts, we get the following:
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

        

1744
1162
17194


## Calculating the Average Number of Comments for Ask HN and Show HN Posts:



In the last screen, we separated the "ask posts" and the "show posts" into two list of lists named ask_posts and show_posts. 

Below are the first five rows in the ask_posts list of lists:
<code>
[

    ['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], 

    ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], 

    ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016', 10:14'], 

    ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], 

    ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']

]
</code>
Below are the first five rows in the show_posts list of lists:
<code>
[

    ['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], 

    ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'], 

    ['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.io', '1', '1', 'h8liu', '4/28/2016 18:05'], 

    ['12178806', 'Show HN: Webscope  Easy way for web developers to communicate with Clients', 'http://webscopeapp.com', '3', '3', 'fastbrick', '7/28/2016 7:11'], 

    ['10872799', 'Show HN: GeoScreenshot  Easily test Geo-IP based web pages', 'https://www.geoscreenshot.com/', '1', '9', 'kpsychwave', '1/9/2016 20:45']

]</code>



Next, let's determine if ask posts or show posts receive more comments on average.

In [9]:
total_ask_comments = 0

for row in ask_posts:
    total_ask_comments += int(row[4])

avg_ask_comments = total_ask_comments / len(ask_posts)

print(avg_ask_comments)

14.038417431192661


In [10]:
total_show_comments = 0

for row in show_posts:
    total_show_comments += int(row[4])

avg_show_comments = total_show_comments / len(show_posts)

print(avg_show_comments)

10.31669535283993


Conclusion: As we can see with our result, we find that there are in fact more ask comments on average (~14.03) than show comments (~10.31).

## Finding the Amount of Ask Posts and Comments by Hour Created

<p>Next, we'll determine if ask posts created at a certain time are more likely to attract comments. We'll use the following steps to perform this analysis:
<ul>
    Calculate the amount of ask posts created in each hour of the day, along with the number of comments received.</ul>
<ul>    Calculate the average number of comments ask posts receive by hour created.</ul>


In this section, we'll tackle the first step — calculating the amount of ask posts and comments by hour created. We'll use <i>datetime</i> to work with the data in the "created_at" column.</p>

We can use the <code>datetime.strptime()</code> constructor to parse dates stored as strings and return datetime objects:

<code>
date_1_str = "December 24, 1984"

date_1_dt = dt.datetime.strptime(date_1_str, "%B %d, %Y")
</code>

Note in the example above that <code>%B</code> represents the full name of the month, <code>%d</code> represents a two digit day, and <code>%Y</code> represents the four digit year. To parse a string in a different format, we will need to make a slight adjustment to the codes above.

Let's use this technique to calculate the amount of ask posts created per hour, along with the total amount of comments.



In [13]:
import datetime as dt

#A list of lists:

result_list = []

for post in ask_posts:
    result_list.append(
        [post[6], int(post[4])]
    )

#Create two dictionaries:
comments_by_hour = {}
counts_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

#Iterate through result_list:
#Asgn. date, comment to row[0], and row[1] 
#Parse the time with the datetime.strptime() function
#Decide if hour is or isnt in key within dictionry "counts_by_hour"
#Display result

for each_row in result_list:
    date = each_row[0]
    comment = each_row[1]
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    if time in counts_by_hour:
        comments_by_hour[time] += comment
        counts_by_hour[time] += 1
    else:
        comments_by_hour[time] = comment
        counts_by_hour[time] = 1

comments_by_hour

{'00': 447,
 '01': 683,
 '02': 1381,
 '03': 421,
 '04': 337,
 '05': 464,
 '06': 397,
 '07': 267,
 '08': 492,
 '09': 251,
 '10': 793,
 '11': 641,
 '12': 687,
 '13': 1253,
 '14': 1416,
 '15': 4477,
 '16': 1814,
 '17': 1146,
 '18': 1439,
 '19': 1188,
 '20': 1722,
 '21': 1745,
 '22': 479,
 '23': 543}

## Calculating the Average Number of Comments for Ask HN Posts by Hour:



In the last screen, we created two dictionaries:

   <code>counts_by_hour</code>: contains the number of ask posts created during each hour of the day.
    <code>comments_by_hour</code>: contains the corresponding number of comments ask posts created at each hour received.

Next, we'll use these two dictionaries to calculate the average number of comments for posts created during each hour of the day.

To illustrate the technique, let's work with the following dictionary:
<code>
sample_dict = {

                'apple': 2, 

                'banana': 4, 

                'orange': 6

               }

</code>
Suppose we wanted to multiply each of the values by ten and return the results as a list of lists. We can use the following code:
<coce>
fruits = []



for fruit in sample_dict:

    fruits.append([fruit, 10*sample_dict[fruit]])

Below are the results:

[['apple', 20], ['banana', 40], ['orange', 60]]

</code>
In the example above, we:

<ul>
    Initialized an empty list (of lists) and assigned it to fruits.</ul>
   <ul> Iterated over the keys of sample_dict and appended to fruits a list whose:</ul>
      <ul>  First element is the key from sample_dict.</ul>
       <ul> Second element is the value corresponding to that key multiplied by ten.</ul>
       
Let's use this format to create a list of lists containing the hours during which posts were created and the average number of comments those posts received.


In [17]:
avg_by_hour = []

for hr in comments_by_hour:
    avg_by_hour.append([hr, comments_by_hour[hr] / counts_by_hour[hr]])

avg_by_hour

[['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['02', 23.810344827586206],
 ['21', 16.009174311926607],
 ['08', 10.25],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297],
 ['09', 5.5777777777777775],
 ['23', 7.985294117647059],
 ['05', 10.08695652173913],
 ['04', 7.170212765957447],
 ['13', 14.741176470588234],
 ['15', 38.5948275862069],
 ['16', 16.796296296296298],
 ['11', 11.051724137931034],
 ['07', 7.852941176470588],
 ['19', 10.8],
 ['17', 11.46],
 ['22', 6.746478873239437],
 ['00', 8.127272727272727],
 ['01', 11.383333333333333],
 ['12', 9.41095890410959],
 ['06', 9.022727272727273],
 ['20', 21.525]]

## Sorting and Printing Values from a List of Lists

<p>In the last screen, we calculated the average number of comments for posts created during each hour of the day, and stored the results in a list of lists named avg_by_hour:</p>

<code>

[

 ['09', 5.5777777777777775],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['17', 11.46],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297]... ]</code>
 
 <p>
Although we now have the results we need, this format makes it hard to identify the hours with the highest values. Let's finish by sorting the list of lists and printing the five highest values in a format that's easier to read.</p>

In [19]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])

print(swap_avg_by_hour)

sorted_swap = sorted(swap_avg_by_hour, reverse = True)
sorted_swap


[[13.440677966101696, '10'], [13.233644859813085, '14'], [23.810344827586206, '02'], [16.009174311926607, '21'], [10.25, '08'], [13.20183486238532, '18'], [7.796296296296297, '03'], [5.5777777777777775, '09'], [7.985294117647059, '23'], [10.08695652173913, '05'], [7.170212765957447, '04'], [14.741176470588234, '13'], [38.5948275862069, '15'], [16.796296296296298, '16'], [11.051724137931034, '11'], [7.852941176470588, '07'], [10.8, '19'], [11.46, '17'], [6.746478873239437, '22'], [8.127272727272727, '00'], [11.383333333333333, '01'], [9.41095890410959, '12'], [9.022727272727273, '06'], [21.525, '20']]


[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [22]:
print("TOP 5 Hours for Ask Posts Comments:")

for avg, hour in sorted_swap[:5]:
    print("{}: {:.2f} average comments per post".format(
            dt.datetime.strptime(hour, "%H").strftime("%H:%M"),avg))




TOP 5 Hours for Ask Posts Comments:
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


## Conclusion:



The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

According to the data set documentation, the timezone used is Eastern Time in the US. So, we could also write 15:00 as 3:00 pm est.


In this project, I analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, I'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.
